### ISSUE TITLE
An error is reported when using metered billing

### ISSUE THREAD

**Issue by dblackherod** [stripe/stripe-firebase-extensions#134]
The latest version (0.19) does not checkout successfully and reports the following in a browser alert box.
**Quantity should not be specified where usage_type is `metered`. Remove quantity from `line_items[0]`**

A peek into the functions/src/index.ts file shows a culprit (ln 125 - 133), although the extension's update message clearly says this version is beta.

I believe a quick fix with an elegant conditional expression should resolve this.

**Comment by thorsten-stripe** [comment]
You can specify the `line_items` array instead of the `price` id to avoid this. See an example of using `line_items` here: https://github.com/stripe/stripe-firebase-extensions/blob/next/firestore-stripe-subscriptions/POSTINSTALL.md#adding-multiple-prices-including-one-time-setup-fees

**Comment by dblackherod** [comment]
@thorsten-stripe 
I do not quite understand what you mean. Perhaps there is some code snippet to illustrate your explanation.
The link in your reply provides a code snippet that still specifies quantity (the culprit variable) while the error encountered clearly says to leave out this variable.

What we would like to achieve is to allow metered billing, and this pricing model cannot function with quantity since there is no way of knowing the quantity beforehand.

I assumed that this stripe-firebase-extension would allow us setup metered billing successfully, but it seems some changes need to be made within the culprit lines in index.ts file to have metered billing function properly with this extension since the metered billing pricing model complains with quantity set in line_items.

I would really appreciate an explanation why the example you provided in your comment will work with the current version 0.19 still declaring quantity in src/index.ts:125-133

**Comment by thorsten-stripe** [comment]
> The link in your reply provides a code snippet that still specifies quantity (the culprit variable) while the error encountered clearly says to leave out this variable.

Yes, and that's exactly what you need to do, remove the quantity. But otherwise the code sample is the same:

```js
const docRef = await db
    .collection('customers')
    .doc(currentUser)
    .collection('checkout_sessions')
    .add({
      line_items: [
        {
          price: 'price_1HCUD4HYgolSBA35icTHEXd5', // metered billing price
        },
      ],
      success_url: window.location.origin,
      cancel_url: window.location.origin,
    });
```

**Comment by dblackherod** [comment]
Hey Thor...
Its great to see real support for awesome and elegant code.
I will try out your snippet and reply on the outcome.
I dare say also that it looks neat you know... like plug n play.

Cheers...

**Comment by dblackherod** [comment]
Hey Thor...
A quick one on this subject...
Do parameters like tax_rates and allow_promotion_codes and meta_data still
work?

I ask since the price parameter is now wrapped in a dictionary which is an
array entry.

Hoping to hear from you soonest.

Regards.

**Comment by thorsten-stripe** [comment]
> Do parameters like tax_rates and allow_promotion_codes and meta_data still
work?

Tax rates have to be provided within the `line_items`: https://github.com/stripe/stripe-firebase-extensions/blob/master/firestore-stripe-subscriptions/POSTINSTALL.md#applying-tax-rates-dynamically

The other parameters (`allow_promotion_codes` and `metadata`) are top-level.

**Comment by dblackherod** [comment]
Hey Thor,
It works like a charm, just like I imagined.

One last thing though... I am lost trying to figure how the usage is
calculated.
I imagined there would be some kind of call to the Usage API to notify when
used.

Am I missing something?

Regards.

**Comment by dblackherod** [comment]
Hey Thor,
I have just read the Stripe Metered Billing
<https://stripe.com/docs/billing/subscriptions/metered> documentation.
It talks about what I mentioned in my previous email that I imagined.

Turns out that usage must be reported
<https://stripe.com/docs/billing/subscriptions/metered#report-usage> during
the billing cycle.
So my question should now be... Does this awesome extension support
reporting usage for metered billing?
How can I achieve this if it does?

The above snippet successfully subscribes user to metered billing model but
how to report usage?

Hoping to hear from you soon.

Best Regards

**Comment by dblackherod** [comment]
Hey Thor,
I am so impressed with this extension so please pardon my intrusion again.

I just want to share with you an idea I have imagined should help with what
I want to achieve.

First off, I want to successfully implement metered billing with Stripe in
my node/express application.

I have used your firebase-subscriptions-payments
<https://github.com/stripe-samples/firebase-subscription-payments>
repository as a starter for my example project.
The code snippet you kindly illustrated worked perfectly by replacing the
appropriate line in the checkout handler method.

Stripe Docs advices that usage reporting should consider incrementing usage
count over a billing period and sending the usage report one time.
This approach should avoid API calls each time the project is used, so I am
trying to figure out how this could work given the current firestore data
structure.

First off, I am thinking of an entry in the customer's subscription
document that indicates usage_frequency since it is payment metadata.
Second, I am thinking a code snippet that increments this
usage_frequency entry (+1) in Firestore when the project (and features) is
used.
Finally, I am thinking of a code snippet that gets the usage_frequency
entry value of customer subscription and reports usage at end of billing
cycle.

In a nutshell, I imagine that your stripe-firebase-extension can define
builtin methods for doing the above 3 steps if billing model is metered.

I also imagined doing all of this without any updates to the extension from
your end, but for now I cannot seemt to figure out where to start.
I know that Step 2 in my imagined flow above can happen anywhere in the
project codex where usage should be reported, but I am pretty sure Step 3
above should be builtin.
I think Step 3 should be builtin since it is the aggregator method that
brings the recommended Stripe approach to completion.

Kindly share your thoughts. If you could be as awesome as recently and
share code illustrations, that'll also be wonderful support.

Hoping to hear from you possibly soonest,

Regards,
Angel

**Comment by thorsten-stripe** [comment]
Documentation on how to report usage can be found here: https://stripe.com/docs/billing/subscriptions/metered-billing#reporting-usage This is something you need to implement yourself as it tends to be very custom depending on your use case and can't be easily standardized.

**Comment by dblackherod** [comment]
Hey Thor,
Yes I understand what you mean.

Did you have a look at my last email?
What do you make of this suggestion?
I was thinking some utility methods for usage reporting would be great.
At least Stripe documentation says to throttle usage reporting and not
report every single usage.
So it would be nice to have utility methods for incrementing usage
frequency and report total usage frequency per billing cycle.

For me, I am curious how to set about achieving this without breaking
functionality in your codebase. That's all.
I could chart a development path from wherever you guidance on this request
leads.

Simply put, if you could help with some semi-complete code then I could get
the job done without breaking too much.

Hoping to hear from you soonest,

Regards,
Angel

**Comment by thorsten-stripe** [comment]
As mentioned, due to the very custom nature of different metered billing use cases we would not add this to the extension. You would write separate Cloud Functions and implement this yourself. If you have questions about your specific implementation please get in touch with our support team or jump on our technical chat on IRC: https://webchat.freenode.net/#stripe

**Comment by dblackherod** [comment]
Hey Thor,
thanks for reaching out.

I'll follow up via those channels.

Regards

**Comment by dblackherod** [comment]
Hey Thor,
Kindly share the support link for the extension so I can get some help
moving forward with our usecase.

Regards.

**Comment by thorsten-stripe** [comment]
https://support.stripe.com/
